internal/backend/runtime/omni/controllers/omni/machine_teardown.go

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) 2024 Sidero Labs, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+package omni
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/controller/generic"
+	"github.com/cosi-project/runtime/pkg/controller/generic/qtransform"
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/gen/optional"
+	"github.com/siderolabs/gen/xerrors"
+	"github.com/siderolabs/talos/pkg/machinery/client"
+	"go.uber.org/zap"
+
+	"github.com/siderolabs/omni/client/pkg/omni/resources"
+	"github.com/siderolabs/omni/client/pkg/omni/resources/omni"
+	"github.com/siderolabs/omni/internal/backend/runtime/talos"
+)
+
+// MachineTeardownControllerName is the name of the MachineTeardownController.
+const MachineTeardownControllerName = "MachineTeardownController"
+
+// MachineTeardownController processes additional teardown steps for a machine leaving a machine set.
+type MachineTeardownController struct {
+	generic.NamedController
+}
+
+// NewMachineTeardownController initializes MachineTeardownController.
+func NewMachineTeardownController() *MachineTeardownController {
+	return &MachineTeardownController{
+		NamedController: generic.NamedController{
+			ControllerName: MachineTeardownControllerName,
+		},
+	}
+}
+
+// Settings implements controller.QController interface.
+func (ctrl *MachineTeardownController) Settings() controller.QSettings {
+	return controller.QSettings{
+		Inputs: []controller.Input{
+			{
+				Namespace: resources.DefaultNamespace,
+				Type:      omni.MachineStatusType,
+				Kind:      controller.InputQPrimary,
+			},
+			{
+				Namespace: resources.DefaultNamespace,
+				Type:      omni.TalosConfigType,
+				Kind:      controller.InputQMapped,
+			},
+		},
+		Concurrency: optional.Some[uint](4),
+	}
+}
+
+// MapInput implements controller.QController interface.
+func (ctrl *MachineTeardownController) MapInput(ctx context.Context, _ *zap.Logger, r controller.QRuntime, ptr resource.Pointer) ([]resource.Pointer, error) {
+	if ptr.Type() == omni.TalosConfigType {
```

Comment:
this is a bit confusing, if it's a one-shot operation, why do we want to reconcile on TalosConfig change?